### PR TITLE
Automate secrets encryption and decryption

### PR DESCRIPTION
**Pull request by Edwin-Pau**
This python script allows the user to automate the encryption and decryption of the terraform.tfvars file so that it is ready to be used by Terraform for both types of deployments.

### CODE REVIEW

File: docs/README-gcp.md [hunk 1]
```diff
@@ -36,14 +36,22 @@ Login to Cloud Access Manager Admin Console at https://cam.teradici.com using a
...
+
+The script rewrites your existing terraform.tfvars secrets using ciphertext so that it is ready to be used for Terraform deployment. In addition, the CAM credentials JSON file will also be replaced with an encrypted file ready to be used by Terraform. The path to the encrypted CAM credentials file will be updated inside of terraform.tfvars.
+
+There are two modes for the script, use the ```-e``` flag for encryption of secrets or ```-d``` flag for decryption of secrets. Decryption mode allows the script to take your terraform.tfvars file containing encrypted secrets and decrypts them back to the original plaintext. The script rewrites the ciphertext version of terraform.tfvars file with the plaintext secrets again.
+
+The script also prompts the user for a deployment type, since there are three different deployment types. Enter ```0``` for single-connector, ```1``` for multi-region or ```1``` for dc-only terraform.tfvars.
```
Comment (shermanyin) [inline docs/README-gcp.md]: 0,1,1?

File: deployments/gcp/single-connector/terraform.tfvars.sample [hunk 2]
```diff
@@ -51,19 +65,29 @@ centos_std_instance_count = 0
 
 centos_admin_ssh_pub_key_file = "~/.ssh/id_rsa.pub"
 
-###############
-#   Secrets   #
-###############
+####################
+#      GCP KMS     #
```
Comment (shermanyin) [inline deployments/gcp/single-connector/terraform.tfvars.sample]: I think leaving this section as "Secrets" is fine, since there is really only 2 actual config lines

File: deployments/gcp/single-connector/terraform.tfvars.sample [hunk 3]
```diff
@@ -51,19 +65,29 @@ centos_std_instance_count = 0
...
+####################
 # The secrets below may be supplied in 2 formats:
 #   1. plain text
 #   2. KMS encrypted, base64 encoded
-# For option 1, leave "kms_cryptokey_id" commented out.
-# For option 2, set "kms_cryptokey_id" to the KMS key used to encrypt the
-# secrets. The GCP service account specified above in "gcp_service_account"
-# must be in the same project and have KMS decryptor permissions for this key.
+# For option 1, leave both keyring name and cryptokey name commented out.
+# For option 2, run the script located under tools using command "python3
```
Comment (shermanyin) [inline deployments/gcp/single-connector/terraform.tfvars.sample]: shouldn't need "python3" in front

File: deployments/gcp/single-connector/terraform.tfvars.sample [hunk 4]
```diff
@@ -51,19 +65,29 @@ centos_std_instance_count = 0
...
+# gcp_kms_encrypto_secrets" which will overwrite this tfvars file with
+# your encrypted secrets and create an encrypted JSON for your CAM
+# credentials file. To set a custom keyring name and cryptokey name,
+# uncomment the two variables below and enter the new names, otherwise
+# a default name will be used by the script. The GCP service account
```
Comment (shermanyin) [inline deployments/gcp/single-connector/terraform.tfvars.sample]: don't need the last 2 sentences

File: deployments/gcp/multi-region/terraform.tfvars.sample [hunk 5]
```diff
@@ -35,6 +44,12 @@ cac_admin_ssh_pub_key_file = "~/.ssh/id_rsa.pub"
 ssl_key  = "/path/to/privkey.pem"
 ssl_cert = "/path/to/fullchain.pem"
 
+
+####################
+#   Workstations   #
+####################
+# Specify the number of workstations created for each type.
```
Comment (shermanyin) [inline deployments/gcp/multi-region/terraform.tfvars.sample]: remove. make sure the formatting are consistent between deployments

File: tools/gcp_kms_secrets.py [hunk 6]
```diff
@@ -0,0 +1,406 @@
...
+
+import argparse
+import base64
+import json
+import os
+from google.cloud        import kms_v1
+from google.cloud.kms_v1 import enums
+from google.oauth2       import service_account
+
+# Global variables
```
Comment (shermanyin) [inline tools/gcp_kms_secrets.py]: remove

File: tools/gcp_kms_secrets.py [hunk 7]
```diff
@@ -0,0 +1,406 @@
...
+
+            lines.append(line.rstrip())
+    
+    # Rewrite the existing terraform.tfvars
+    with open(tfvars_file, 'w') as f:
+        f.writelines("%s\n" %line for line in lines)
+
+# Use argparse to determine user's desired action, encryption or decryption of terraform.tfvars
+def main():
+    parser = argparse.ArgumentParser(description="This script encrypts or decrypts the secrets  \
```
Comment (shermanyin) [inline tools/gcp_kms_secrets.py]: If we have a long string like this, I typically put it in a var and then just reference the var here instead of creating a multi-line function call. see the quickstart wrapper script.

File: tools/gcp_kms_secrets.py [hunk 8]
```diff
@@ -0,0 +1,406 @@
...
+                                    inside the user-specified terraform.tfstate file by         \
+                                    prompting the user for one of the three deployment types    \
+                                    (single, multi-region, dc-only). Select the mode by setting \
+                                    a flag: -e for encryption or -d for decryption")
```
Comment (shermanyin) [inline tools/gcp_kms_secrets.py]: I'm thinking it's probably more flexible to let the user specify which file they want to run the tool on instead of making them select 1, 2 or 3 and hard code the paths.  It's bad enough that we're hard-coding which values to encrypt now.  I'm debating whether we should move all the secrets to a separate tfvars file.  

So a user might run "tfvars-encrypt ../deployments/gcp/single-connector/terrafrom.tfvars" 

Suggestion: "Creates GCP KMS keyring and key, and use the key to encrypt secrets in terraform.tfvars specified. "

File: tools/gcp_kms_secrets.py [hunk 9]
```diff
@@ -0,0 +1,406 @@
...
+
+    parser.add_argument("-e", help="run this script in encryption mode", action='store_true')
```
Comment (shermanyin) [inline tools/gcp_kms_secrets.py]: There should be no flag needed for encrypt, which is the default. just -d for decrypt

File: tools/gcp_kms_secrets.py [hunk 10]
```diff
@@ -0,0 +1,406 @@
...
+
+    args = parser.parse_args()
+    
+    if args.e:
+        print("Encryption mode selected...\n")
+        encrypt_tfvars_secrets()
+    elif args.d:
+        print("Decryption mode selected...\n")
+        decrypt_tfvars_secrets()
+    else:
```
Comment (shermanyin) [inline tools/gcp_kms_secrets.py]: this would not be needed if the default is -e.

File: tools/gcp_kms_secrets.py [hunk 11]
```diff
@@ -0,0 +1,406 @@
+#!/usr/bin/python3
```
Comment (shermanyin) [inline tools/gcp_kms_secrets.py]: I know I don't do this either and I should, but read up on https://www.python.org/dev/peps/pep-0257/ and document your functions properly.

Comment (shermanyin) [inline tools/gcp_kms_secrets.py]: A bit too many comments. I generally like comments as it explains clearly the author's intentions, but if it is very obvious then there doesn't need to be comments. e.g.
    # Create a key ring
    response = client.create_key_ring(parent, key_ring_id, keyring)
Also, try to document why you are doing something rather than what you are doing (unless it's not obvious). Like the above, you are obviously creating a key ring by calling "create_key_ring()"

File: tools/gcp_kms_secrets.py [hunk 12]
```diff
@@ -0,0 +1,406 @@
...
+    tf_data = {}
+
+    with open(tfvars_file, 'r') as f:
+        for line in f:
+            if line[0] in ('#', '\n'):
+                continue
+            
+            # Split using the first delimiter
+            key, value = map(str.strip, line.split('=', 1))
+            tf_data[key] = value.replace("\"", "")
```
Comment (shermanyin) [inline tools/gcp_kms_secrets.py]: where do we run into this issue of having \"?

File: tools/gcp_kms_secrets.py [hunk 13]
```diff
@@ -0,0 +1,406 @@
...
+
+# Collect the required secrets for the appropriate deployment from the tfvars file
+def get_tfvars_secrets(tfvars_dict, tfvars_path):
+    secrets = {
+        "dc_admin_password"           : tfvars_dict.get("dc_admin_password"),
+        "safe_mode_admin_password"    : tfvars_dict.get("safe_mode_admin_password"),
+        "ad_service_account_password" : tfvars_dict.get("ad_service_account_password")
+    }
+    
+    if "single-connector" in tfvars_path or "multi-region" in tfvars_path:
```
Comment (shermanyin) [inline tools/gcp_kms_secrets.py]: don't think we should rely on the path.

how about just encrypt fields as we find them, and ignore if they are missing.

File: tools/gcp_kms_secrets.py [hunk 14]
```diff
@@ -0,0 +1,406 @@
...
+    credentials = service_account.Credentials.from_service_account_file(GCP_CREDENTIALS_FILE)
+    client      = kms_v1.KeyManagementServiceClient(credentials = credentials)
+
+    # GCP KMS resource variables
+    project_id = tfvars_dict.get("gcp_project_id")
+    location   = "global"
+
+    if tfvars_dict.get("kms_keyring_name")   == None or \
+       tfvars_dict.get("kms_cryptokey_name") == None:
+        key_ring_id   = "terraform-keyring"
```
Comment (shermanyin) [inline tools/gcp_kms_secrets.py]: this kind of breaks the convention of adding the prefix to all resources created.

File: tools/gcp_kms_secrets.py [hunk 15]
```diff
@@ -0,0 +1,406 @@
...
+
+    # Create an API client for the KMS API using the provided GCP service account
+    credentials = service_account.Credentials.from_service_account_file(GCP_CREDENTIALS_FILE)
+    client      = kms_v1.KeyManagementServiceClient(credentials = credentials)
+
+    # GCP KMS resource variables
+    project_id = tfvars_dict.get("gcp_project_id")
+    location   = "global"
+
+    if tfvars_dict.get("kms_keyring_name")   == None or \
```
Comment (shermanyin) [inline tools/gcp_kms_secrets.py]: I'm thinking maybe we don't even let users choose the name. just call them cas_keyring and cas_key or something.

File: tools/gcp_kms_secrets.py [hunk 16]
```diff
@@ -0,0 +1,406 @@
...
+       tfvars_dict.get("kms_cryptokey_name") == None:
+        key_ring_id   = "terraform-keyring"
+        crypto_key_id = "terraform-cryptokey"
+    else:
+        key_ring_id   = tfvars_dict.get("kms_keyring_name")
+        crypto_key_id = tfvars_dict.get("kms_cryptokey_name")
+
+    # List all key rings
+    list_key_rings(client, project_id, location)
```
Comment (shermanyin) [inline tools/gcp_kms_secrets.py]: why do we need this?

File: tools/gcp_kms_secrets.py [hunk 17]
```diff
@@ -0,0 +1,406 @@
...
+
+    # Create the key ring
+    try:
+        kms_keyring_id = create_key_ring(client, project_id, location, key_ring_id)
+        print("Created key ring: {}\n".format(kms_keyring_id))
+    
+    except Exception as err:
```
Comment (shermanyin) [inline tools/gcp_kms_secrets.py]: not sure what the logic here is. you're trying to take care of the case when the user-specified key ring and key already exist? If so I prefer if you explicitly check for it rather than try and catch the error, because there could be other errors other than "key already exist".

File: tools/gcp_kms_secrets.py [hunk 18]
```diff
@@ -0,0 +1,406 @@
...
+        print("An exception occurred creating new crypto key:")
+        print("{}".format(err))
+        print("Using existing crypto key: {}\n".format(crypto_key_id))
+
+    # Encrypt secrets including CAM credentials
+    try:
+        for secret in secrets:
+            print("Encrypting {}...\n".format(secret))
+            ciphertext = encrypt_plaintext(client, project_id, location, key_ring_id, crypto_key_id, secrets.get(secret))
+            print("Finished encrypting {0}:\n{1}\n".format(secret, ciphertext))
```
Comment (shermanyin) [inline tools/gcp_kms_secrets.py]: don't need 0 and 1.  probably no need to print out the ciphertext either since we don't expect them to be legible or needs copying.

File: tools/gcp_kms_secrets.py [hunk 19]
```diff
@@ -0,0 +1,406 @@
...
+            secrets[secret] = ciphertext
+
+        if tfvars_dict.get("cam_credentials_file"):
+            encrypt_cam_credentials(client, project_id, location, key_ring_id, crypto_key_id)
+            
+    except Exception as err:
+        print("An exception occurred encrypting secrets:")
+        print("{}\n".format(err))
```
Comment (shermanyin) [inline tools/gcp_kms_secrets.py]: if an exception occur, shouldn't you exit?

File: tools/gcp_kms_secrets.py [hunk 20]
```diff
@@ -0,0 +1,406 @@
...
+            # Split using the first delimiter
+            key, value = map(str.strip, line.split('=', 1))
+            tf_data[key] = value.replace("\"", "")
+
+    return tf_data
+
+
+# Write the new tfvars file using encrypted secrets
+def write_new_tfvars(tfvars_file, secrets, kms_cryptokey_id):
+    kms_resource_list  = kms_cryptokey_id.split('/')
```
Comment (shermanyin) [inline tools/gcp_kms_secrets.py]: best if you can give an example of why "5" and "7" are chosen, so reader doesn't have to google what the string is supposed to look like.

File: tools/gcp_kms_secrets.py [hunk 21]
```diff
@@ -0,0 +1,406 @@
...
+    kms_keyring_name   = kms_resource_list[5]
+    kms_cryptokey_name = kms_resource_list[7]
+
+    lines = []
+    
+    with open(tfvars_file, 'r') as f:
+        for line in f:
+            if "dc_admin_password" in line:
```
Comment (shermanyin) [inline tools/gcp_kms_secrets.py]: what if I have a comment line with "dc_admin_password" in it somewhere?

File: tools/gcp_kms_secrets.py [hunk 22]
```diff
@@ -0,0 +1,406 @@
...
+            
+            # Split using the first delimiter
+            key, value = map(str.strip, line.split('=', 1))
+            tf_data[key] = value.replace("\"", "")
+
+    return tf_data
+
+
+# Write the new tfvars file using encrypted secrets
+def write_new_tfvars(tfvars_file, secrets, kms_cryptokey_id):
```
Comment (shermanyin) [inline tools/gcp_kms_secrets.py]: need to save the original as a backup.  Log this file name to the output to remind the user where the backup is.

**Review by shermanyin — CHANGES_REQUESTED** [review]

File: tools/gcp_kms_secrets.py (hunk 12 above)
Comment (Edwin-Pau) [reply tools/gcp_kms_secrets.py]: The above block of the code was my fix to the double quotes problem that I had a while back.

When I reused the quickstart_config_read function in your quickstart.py to read in lines, I made the mistake of using it without modifying it for this use (so it was grabbing the double quotes in tfvars secrets and treating it as part of the secrets when encrypting). 
`key, value = map(str.strip, line.split(':'))`
`cfg_data[key] = value`
I didn't think much of it the first time since I sanitized the double quotes during decryption output so I didn't think this was the issue. 

Lesson learned!

File: tools/gcp_kms_secrets.py (hunk 15 above)
Comment (Edwin-Pau) [reply tools/gcp_kms_secrets.py]: Going with your suggestion of just creating the keyring and key for a better user experience, do I still want to add the user-specified prefix (if they used the field in terraform.tfvars) to this?

File: tools/gcp_kms_secrets.py (hunk 22 above)
Comment (Edwin-Pau) [reply tools/gcp_kms_secrets.py]: Do we still want the decryption function if we are saving original plaintext as backup?

File: tools/gcp_kms_secrets.py (hunk 6 above)
Comment (Edwin-Pau) [reply tools/gcp_kms_secrets.py]: I believe I need the service_account module to create credentials used for a client:
`credentials = service_account.Credentials.from_service_account_file(GCP_CREDENTIALS_FILE)`

`client = kms_v1.KeyManagementServiceClient(credentials = credentials)`

File: tools/gcp_kms_secrets.py (hunk 22 above)
Comment (shermanyin) [reply tools/gcp_kms_secrets.py]: Good point :)  I guess it's not really needed if we have a backup.

File: docs/README-gcp.md [hunk 23]
```diff
@@ -38,12 +38,22 @@ Login to Cloud Access Manager Admin Console at https://cam.teradici.com using a
...
 
-The Terraform scripts are designed to support both plaintext and KMS-encrypted secrets. Plaintext secrets requires no extra steps, but will be stored in plaintext in the above mentioned locations. It is recommended that secrets are first encrypted before being entered into terraform.tfvars.
+The Terraform scripts are designed to support both plaintext and KMS-encrypted secrets. Plaintext secrets requires no extra steps, but will be stored in plaintext in the above mentioned locations. It is recommended to encrypt the secrets in the terraform.tfvars file before deploying. Secrets can be encrypted manually first before being entered into terraform.tfvars, or they can be encrypted using a python script located under tools.
 
 To encrypt secrets using the KMS crypto key created above, follow the instructions here: https://cloud.google.com/kms/docs/encrypt-decrypt. Base64 encode the ciphertext before copying and pasting it into terraform.tfvars. For example, execute the following command in GCP Cloud Shell:
```
Comment (shermanyin) [inline docs/README-gcp.md:44]: we can add a section header here called "Manual encryption"

File: docs/README-gcp.md [hunk 24]
```diff
@@ -38,12 +38,22 @@ Login to Cloud Access Manager Admin Console at https://cam.teradici.com using a
...
 
 ```echo -n <secret> | gcloud kms encrypt --location <location> --keyring <keyring_name> --key <key_name> --plaintext-file - --ciphertext-file - | base64```
 
+Alternatively, run the ```gcp_kms_secrets``` python3 script under the tools directory to automate the KMS encryption process. This script creates a key ring and crypto key on GCP KMS, then it encrypts all the secrets inside of your terraform.tfvars using this key. It will also encrypt any files that are entered as a path under the secrets line.
```
Comment (shermanyin) [inline docs/README-gcp.md]: we can add a section header called "Encryption using Python script"

File: tools/kms_secrets_encryption.py [hunk 25]
```diff
@@ -0,0 +1,548 @@
...
+        self.tfvars_path    = tfvars_path
+        self.tfvars_data    = self.read_tfvars_data(tfvars_path)
+        self.tfvars_secrets = self.read_tfvars_secrets(tfvars_path)
+
+        # Set GCP credentials instance variable from tfvars_data
+        self.gcp_credentials_file = self.tfvars_data.get("gcp_credentials_file")
+
+        # Create an API client for the KMS API using the provided GCP service account
+        self.credentials = service_account.Credentials.from_service_account_file(self.gcp_credentials_file)
+        self.client      = kms_v1.KeyManagementServiceClient(credentials = self.credentials)
```
Comment (shermanyin) [inline tools/kms_secrets_encryption.py]: i would call this kms_client

File: tools/kms_secrets_encryption.py [hunk 26]
```diff
@@ -0,0 +1,548 @@
...
+        # Read tfvars data and secrets into dictionaries
+        self.tfvars_path    = tfvars_path
+        self.tfvars_data    = self.read_tfvars_data(tfvars_path)
+        self.tfvars_secrets = self.read_tfvars_secrets(tfvars_path)
+
+        # Set GCP credentials instance variable from tfvars_data
+        self.gcp_credentials_file = self.tfvars_data.get("gcp_credentials_file")
+
+        # Create an API client for the KMS API using the provided GCP service account
+        self.credentials = service_account.Credentials.from_service_account_file(self.gcp_credentials_file)
```
Comment (shermanyin) [inline tools/kms_secrets_encryption.py]: i would call it gcp_credentials

File: tools/kms_secrets_encryption.py [hunk 27]
```diff
@@ -0,0 +1,548 @@
...
+        self.client      = kms_v1.KeyManagementServiceClient(credentials = self.credentials)
+
+        # GCP KMS resource variables
+        self.project_id      = self.tfvars_data.get("gcp_project_id")
+        self.location        = "global"
```
Comment (shermanyin) [inline tools/kms_secrets_encryption.py:78]: If it's just going to be a static "global", just make it a global variable at the top of the file and refer to that.

File: tools/kms_secrets_encryption.py [hunk 28]
```diff
@@ -0,0 +1,548 @@
...
+
+import argparse
+import base64
+import json
+import os
+from google.cloud  import kms_v1
+from google.oauth2 import service_account
+
+
+class GCP_KMS_Client:
```
Comment (shermanyin) [inline tools/kms_secrets_encryption.py]: gcp_kms_client sounds like what google.cloud.kms_v1 returns.  I'd call this something like tfvar_gcp_encryptor or something.

File: tools/kms_secrets_encryption.py [hunk 29]
```diff
@@ -0,0 +1,548 @@
...
+        with the GCP service account.
+        
+        Returns:
+            list: a list of all the key rings
+        """
+
+        parent   = self.client.location_path(self.project_id, self.location)
+        response = self.client.list_key_rings(parent)
+
+        # Extract just the key ring name from each response and put into a list
```
Comment (shermanyin) [inline tools/kms_secrets_encryption.py]: it's often helpful to give an examle of what the the output from list_key_rings() so that we know what you are trying to process.  e.g. "the response returns something like this: "some/path/keyname" and we're extracting the "keyname" only".

File: tools/kms_secrets_encryption.py [hunk 30]
```diff
@@ -0,0 +1,548 @@
...
+        Args:
+            key_ring_id (str): key ring being created
+        Returns:
+            string: the key ring used
+        """
+
+        key_rings_list = self.get_key_rings()
+
+        # Create the key ring only if it doesn't exist
+        if key_ring_id not in key_rings_list:
```
Comment (shermanyin) [inline tools/kms_secrets_encryption.py:340]: Good that you are checking if the key ring already exist.  I would print a message for the case when an existing key ring is used instead. "Using existing keyring {name}"

Also, a more pythonic way to do this is go ahead and try to create the key ring without checking, then presumably GCP will error and tell you the ring exist, which you should catch, print "using existing keyring {name}", then continue.  In that case, you can do all that in create_key_ring() so you don't need initialize_keyring().  You don't have to change this but just FYI

File: tools/kms_secrets_encryption.py [hunk 31]
```diff
@@ -0,0 +1,548 @@
...
+
+        crypto_keys_list = []
+
+        for key_ring in key_rings_list:
+            parent   = self.client.key_ring_path(self.project_id, self.location, key_ring)
+            response = self.client.list_crypto_keys(parent)
+
+            # Extract just the crypto key name from each response and put that into a temp list
+            # [2] to get the string after the separator
+            temp_list = list(map(lambda key: key.name.rpartition('/')[2], response))
```
Comment (shermanyin) [inline tools/kms_secrets_encryption.py]: would be nice to include example of what "key" looks like before in the comments.

File: tools/kms_secrets_encryption.py [hunk 32]
```diff
@@ -0,0 +1,548 @@
...
+        create a default crypto key if it doesn't exist.
+        
+        Args:
+            crypto_key_id (str): crypto key used to encrypt and decrypt
+        Returns:
+            string: the crypto key used
+        """
+
+        key_rings_list   = self.get_key_rings()
+        crypto_keys_list = self.get_crypto_keys(key_rings_list)
```
Comment (shermanyin) [inline tools/kms_secrets_encryption.py]: why do we have to create a list of all keys from all key_rings?  aren't we only interested in keys within the same keyring?

File: tools/kms_secrets_encryption.py [hunk 33]
```diff
@@ -0,0 +1,548 @@
...
+
+        # Create a crypto key only if it doesn't exist
+        if crypto_key_id not in crypto_keys_list:
```
Comment (shermanyin) [inline tools/kms_secrets_encryption.py:310]: see comment on line 366. Same applies here

File: tools/kms_secrets_encryption.py [hunk 34]
```diff
@@ -0,0 +1,548 @@
...
+        """
+
+        tf_data = {}
+
+        with open(tfvars_file, 'r') as f:
+            for line in f:
+                if "# Secrets below can be encrypted using GCP-KMS, refer to the README for more information." in line:
+                    break
+
+                if line[0] in ('#', '\n'):
```
Comment (shermanyin) [inline tools/kms_secrets_encryption.py]: i would trim the line first, so that this also works if there are spaces before the first #

File: tools/kms_secrets_encryption.py [hunk 35]
```diff
@@ -0,0 +1,548 @@
...
+            tfvars_file (str): a path to a terraform.tfvars file
+        Returns:
+            dict: key value pairs for all the terraform.tfvars data
+        """
+
+        tf_data = {}
+
+        with open(tfvars_file, 'r') as f:
+            for line in f:
+                if "# Secrets below can be encrypted using GCP-KMS, refer to the README for more information." in line:
```
Comment (shermanyin) [inline tools/kms_secrets_encryption.py]: hmm. can we have something shorter to indicate the separation? if someone accidentally added a space somewhere in the line then it's not going to work. Maybe something like # <--start of secrets section, do not edit this line. --> or something like that.

Also, whatever line you use, make it a global at the top of the file, since u'll use it more than once.

File: tools/kms_secrets_encryption.py [hunk 36]
```diff
@@ -0,0 +1,548 @@
...
+
+        try:
+            print("Decrypting file: {}...".format(file_path))
+
+            with open(file_path) as f:
+                file_ciphertext = f.read()
+        
+            file_decrypted = self.decrypt_ciphertext(file_ciphertext)
+
+            file_path_decrypted = "{}".format(file_path).replace(".encrypted", "")
```
Comment (shermanyin) [inline tools/kms_secrets_encryption.py]: what if the filename doesn't have ".encrypted" in it?

File: tools/kms_secrets_encryption.py [hunk 37]
```diff
@@ -0,0 +1,548 @@
...
+        symmetric crypto key that belongs to this instance.
+        
+        Args:
+            ciphertext (str): the plainttext being encrypted
+        Returns:
+            string: the ciphertext
+        """
+
+        # UTF-8 encoding of plaintext
+        plaintext = plaintext.encode("utf-8")
```
Comment (shermanyin) [inline tools/kms_secrets_encryption.py]: this could be rolled into 274, no need to redefine plaintext

File: tools/kms_secrets_encryption.py [hunk 38]
```diff
@@ -0,0 +1,548 @@
...
+
+        # Use the KMS API to encrypt the data.
+        response = self.client.encrypt(self.crypto_key_path, plaintext)
+
+        # Base64 encoding of ciphertext
+        ciphertext = base64.b64encode(response.ciphertext).decode()
```
Comment (shermanyin) [inline tools/kms_secrets_encryption.py]: do we need to do decode('utf-8')?

File: tools/kms_secrets_encryption.py [hunk 39]
```diff
@@ -0,0 +1,548 @@
...
+            file_path (str): the path of the text file being encrypted
+        Returns:
+            string: the path to the encrypted text file created
+        """
+
+        try:
+            print("Encrypting file: {}...".format(file_path))
+            
+            with open(file_path) as f:
+                f = json.load(f)
```
Comment (shermanyin) [inline tools/kms_secrets_encryption.py]: why do we need to do json.load() then json.dumps() after?

File: tools/kms_secrets_encryption.py [hunk 40]
```diff
@@ -0,0 +1,548 @@
...
+        
+            f_string = json.dumps(f)
+            f_encrypted_string = self.encrypt_plaintext(f_string)
+            file_path_encrypted = "{}.encrypted".format(file_path)
```
Comment (shermanyin) [inline tools/kms_secrets_encryption.py]: Here is what I think you are doing. On encrypting a file, say "a.txt", you encrypt it and save it as "a.txt.encrypted", then rename "a.txt" to "a.txt.backup".  On decrypting, you expect the file to end in ".encrypted", say "b.txt.encrypted", you decrypt it and take out the .encrypted to get "b.txt", then you delete "b.txt.encrypted".
how about:
on encryption: take a.txt, encrypt it and call it a.txt.encrypted. leave a.txt alone
on decryption: take b.txt, decrypt it and call it b.txt.decrypted. leave b.txt alone

File: tools/kms_secrets_encryption.py [hunk 41]
```diff
@@ -0,0 +1,548 @@
...
+        tf_secrets = {}
+        begin_reading_secrets = False
+
+        with open(tfvars_file, 'r') as f:
+            for line in f:
+                if "# Secrets below can be encrypted using GCP-KMS, refer to the README for more information." in line:
+                    begin_reading_secrets = True
+                    continue
+                
+                if line[0] in ('#', '\n'):
```
Comment (shermanyin) [inline tools/kms_secrets_encryption.py]: trim spaces

File: tools/kms_secrets_encryption.py [hunk 42]
```diff
@@ -0,0 +1,548 @@
...
+        with open(self.tfvars_path, 'r') as f:
+            for line in f:
+                # Skip copying kms_cryptokey_id to tfvars if it is decrypting
+                if "kms_cryptokey_id" in line:
+                    continue
+
+                lines.append(line.rstrip())
+
+                # Append kms_cryptokey_id to tfvars if it is encrypting
+                if "# must be in the same project and have KMS decryptor permissions for this key." in line:
```
Comment (shermanyin) [inline tools/kms_secrets_encryption.py]: dun think it's a good idea to check for such a specific string. any updates to the comments will break this

File: tools/kms_secrets_encryption.py [hunk 43]
```diff
@@ -0,0 +1,548 @@
...
+                    if not self.tfvars_data.get("kms_cryptokey_id"):
+                        lines.append("{} = \"{}\"".format("kms_cryptokey_id", self.crypto_key_path))
+
+                # Begin copying all key value pairs in secrets dictionary after this line
+                if "# Secrets below can be encrypted using GCP-KMS, refer to the README for more information." in line:
+                    break
+            
+            # Add all secrets back into tfvars
+            for key, value in self.tfvars_secrets.items():
```
Comment (shermanyin) [inline tools/kms_secrets_encryption.py]: wouldn't this change the order of the key value items? I'd try to keep the order the same. e.g. read each line and parse the key, if it needs updating then replace the value with the encrypted version. Reference how it was done in gcp-cloudshell-quickstart.py line 246

File: tools/kms_secrets_encryption.py [hunk 44]
```diff
@@ -0,0 +1,548 @@
...
+        if not self.tfvars_data.get("gcp_credentials_file"):
+            print("Missing gcp_credentials_file in tfvars. Ensure gcp_credentials_file is valid and try again.\n")
+            raise SystemExit()
+
+        # Encrypt all secrets found in the tfvars_secrets dictionary
+        try:
+            for secret in self.tfvars_secrets:
+                # Additional handling needed if the string is a path to a file (IE. cam_credentials_file)
+                if os.path.isfile(self.tfvars_secrets.get(secret)):
+                    file_encrypted = self.encrypt_file(self.tfvars_secrets.get(secret))
```
Comment (shermanyin) [inline tools/kms_secrets_encryption.py]: dun really need the file_encrypted variable

File: tools/kms_secrets_encryption.py [hunk 45]
```diff
@@ -0,0 +1,548 @@
...
+                    self.tfvars_secrets[secret] = file_encrypted
+                else:
+                    print("Encrypting {}...".format(secret))
+                    ciphertext = self.encrypt_plaintext(self.tfvars_secrets.get(secret))
```
Comment (shermanyin) [inline tools/kms_secrets_encryption.py]: ciphertext not needed.

File: tools/kms_secrets_encryption.py [hunk 46]
```diff
@@ -0,0 +1,548 @@
...
+        write_new_tfvars() to write all secrets to a new terraform.tfvars file. 
+        """    
+
+        # Abort the decryption if the tfvars is in plaintext using kms_cryptokey_id as an indicator
+        if not self.tfvars_data.get("kms_cryptokey_id"):
+            print("No kms_cryptokey_id present in tfvars. Ensure the secrets are encrypted and try again.\n")
+            raise SystemExit()
+
+        # Abort the decryption if GCP credentials is missing
+        if not self.tfvars_data.get("gcp_credentials_file"):
```
Comment (shermanyin) [inline tools/kms_secrets_encryption.py]: could move this check out to the parent since both encrypt and decrypt function calls this.

**Review by shermanyin — CHANGES_REQUESTED** [review]

File: tools/kms_secrets_encryption.py (hunk 27 above)
Comment (Edwin-Pau) [reply tools/kms_secrets_encryption.py:78]: As per my explanation over Slack, I will be keeping this within the GCP instance since AWS in the future will not be using this location variable.

File: tools/kms_secrets_encryption.py (hunk 39 above)
Comment (Edwin-Pau) [reply tools/kms_secrets_encryption.py]: It didn't work if I only use dumps since it requires a JSON object, not a file object. I believe I need to use json.load() to convert a file object to json object, then use .dumps() to convert a json object to a string.
Error if I remove json.load():
`An exception occurred encrypting the file:
Object of type 'TextIOWrapper' is not JSON serializable`

File: docs/README-gcp.md [hunk 47]
```diff
@@ -38,12 +38,23 @@ Login to Cloud Access Manager Admin Console at https://cam.teradici.com using a
...
+The Terraform scripts are designed to support both plaintext and KMS-encrypted secrets. Plaintext secrets requires no extra steps, but will be stored in plaintext in the above mentioned locations. It is recommended to encrypt the secrets in the terraform.tfvars file before deploying. Secrets can be encrypted manually first before being entered into terraform.tfvars, or they can be encrypted using a python script located under tools.
 
+#### Manual Encryption
 To encrypt secrets using the KMS crypto key created above, follow the instructions here: https://cloud.google.com/kms/docs/encrypt-decrypt. Base64 encode the ciphertext before copying and pasting it into terraform.tfvars. For example, execute the following command in GCP Cloud Shell:
 
 ```echo -n <secret> | gcloud kms encrypt --location <location> --keyring <keyring_name> --key <key_name> --plaintext-file - --ciphertext-file - | base64```
 
+#### Encryption Using Python Script
+Alternatively, run the kms_secrets_encryption.py Python 3 script under the tools directory to automate the KMS encryption process.
```
Comment (shermanyin) [inline docs/README-gcp.md:49]: need a step here.  "first, enter the plaintext secrets into terraform.tfvars, save the file, then run the script"

File: docs/README-gcp.md [hunk 48]
```diff
@@ -38,12 +38,23 @@ Login to Cloud Access Manager Admin Console at https://cam.teradici.com using a
...
 
+#### Encryption Using Python Script
+Alternatively, run the kms_secrets_encryption.py Python 3 script under the tools directory to automate the KMS encryption process. 
+
+Execute the following command inside the tools directory:
+```./kms_secrets_encryption.py </path/to/terraform.tfvars>```
+
+The script rewrites your existing terraform.tfvars secrets using ciphertext so that it is ready to be used for your Terraform deployment. In addition, any text files specified under the secrets as a path will also be encrypted so that it is ready to be used by Terraform. The script will also update the path to the encrypted text file inside of the new terraform.tfvars.
+
+The default mode is encryption; to decrypt the secrets back to plaintext, use the ```-d``` flag when executing the script. Decryption mode decrypts the ciphertext in the terraform.tfvars file back to the original plaintext. The script rewrites the ciphertext version of terraform.tfvars file with the plaintext secrets again.
```
Comment (shermanyin) [inline docs/README-gcp.md:56]: A bit wordy.

The script can also reverse the encryption by running with the '-d' flag.  See script's documentation for details (--help)

File: tools/README.md [hunk 49]
```diff
@@ -11,3 +11,25 @@ pip install names
...
+line in terraform.tfvars is commented out.
+
+To encrypt:
+```
+./kms_secrets_encryption.py <path/to/terraform.tfvars>
+```
+
+To decrypt:
+```
+./kms_secrets_encryption.py <path/to/terraform.tfvars> -d
```
Comment (shermanyin) [inline tools/README.md:34]: usually the -d comes before the filename.

File: tools/kms_secrets_encryption.py [hunk 50]
```diff
@@ -0,0 +1,551 @@
...
+        response = self.kms_client.list_key_rings(parent)
+
+        # Access the name property and split string from the right. [2] to get the string after the separator
+        # eg. name: "projects/user-terraform/locations/global/keyRings/cas_keyring"
+        key_rings_list = list(map(lambda key_ring: key_ring.name.rpartition('/')[2], response))
+
+        return key_rings_list
+
+
+    def read_tfvars_data(self, tfvars_file):
```
Comment (shermanyin) [inline tools/kms_secrets_encryption.py:396]: This and the secret reading function could be 1 function since they share so much in common.

**Review by shermanyin — APPROVED** [review]
